src/discovery/local-discovery.js

```diff
@@ -107,17 +111,24 @@ export class LocalDiscovery extends TypedEmitter {
   connectPeer({ address, port, name }) {
     if (this.#name === name) return
     this.#log('peer connected', name.slice(0, 7), address, port)
-    if (this.#noiseConnections.has(name)) {
+    if (this.#connectionsToName.has(name)) {
       this.#log(`Already connected to ${name.slice(0, 7)}`)
       return
     }
+    const ipId = `${address}:${port}`
+    if (this.#connectionsFromIP.has(ipId)) {
+      this.#log(`Already connected to ${ipId}`)
+      return
+    }
     const socket = net.connect({
       ...TCP_KEEP_ALIVE_OPTIONS,
       host: address,
       port,
     })
     socket.on('error', this.#handleSocketError)
     socket.once('connect', () => {
+      this.#connectionsToName.add(name)
```

Comment:
I think we shouldn't add this until the noise handshake and deduplication is complete, to avoid a possible race condition with the deduplication logic where one side dedupes here (first connection wins) and the other side dedupes with the dedupe code after noise stream connection (1st connection sorted by public key wins). This could happen with unusual network conditions, would be rare, but extremely hard to find and debug.